File: search/leetcode_Search_in_Rotated_Sorted_Array.cpp
/*
 * Copyright 2016
 * author: madarfacar
 *
 * leetcode search in rotated sorted array
 */


#include <stdio.h>
#include <vector>

using std::vector;


class Solution {
 public:
  /**
   * The approach is based on a variation of binary searhc.
   *
   * The problem is the given array is rotated. Need to check whether
   * array[head:mid] is sorted or array[mid:tail] is sorted. It's impossible
   * that both array[head:mid] and array[mid:tail] is not sorted(why?). Once
   * the sorted range is found, it's quite easy to tell whether target is in
   * this range or not.
   */
  int search(vector<int>& nums, int target) {
    int head = 0;
    int tail = nums.size() - 1;
    int idx = -1;
    bool found = false;

    while (!found && head <= tail) {
      int mid = (tail - head) / 2 + head;
      if (nums.at(mid) == target) {
        found = true;
        idx = mid;
      } else if (nums.at(head) <= nums.at(mid)) {
        // nums[head:mid] is sorted.
        if (nums.at(head) <= target && target <= nums.at(mid)) {
          tail = mid - 1;
        } else {
          head = mid + 1;
        }
      } else {
        // nums[mid:tail] is sorted.
        if (nums.at(mid) <= target && target <= nums.at(tail)) {
          head = mid + 1;
        } else {
          tail = mid - 1;
        }
      }
    }

    return idx;
  }
};


int main() {
  Solution sol;

  vector<int> nums {4, 5, 6, 7, 8, 1, 2, 3};

  for (int target : nums) {
    int res = sol.search(nums, target);
    printf("target: %d index: %d\n", target, res);
  }
}

File: search/leetcode_Search_in_Rotated_Sorted_Array.java
/*Search_in_Rotated_Sorted_Array

Suppose a sorted array is rotated at some pivot unknown to you beforehand.
(i.e., [0 1 2 4 5 6 7] might become [4 5 6 7 0 1 2]).
You are given a target value to search. If found in the array return its index,
otherwise return -1.
You may assume no duplicate exists in the array.
*/


public class leetcode_Search_in_Rotated_Sorted_Array {

    public static void main(String[] args) {

    }

    public static int search(int[] A, int target) {
        int start = 0, end = A.length - 1, middle;
        while (start <= end) {
			middle = (end - start) / 2 + start;
            if (target == A[middle]) {
                return middle;
            } else if (A[start] <= A[middle]) {
                // [3 4 5 6 7 0 1 2] middle 6 > start 3
                if (A[start] <= target && target <= A[middle]) {
                    end = middle - 1;
                } else {
                    start = middle + 1;
                }
            } else {
                // A[start] > A[middle] [6 7 0 1 2 3 4 5]
                if (A[middle] <= target && target <= A[end]) {
                    start = middle + 1;
                } else {
                    end = middle - 1;
                }
            }
        }
        return -1;
    }
}

/* Python Version
def search(self, A, target):
    if not A:
        return -1
    
    head = 0
    tail = len(A)- 1
    
    while head <= tail:
        mid = head + (tail - head) / 2
        if A[mid] == target:
            return mid
        elif A[head] <= A[mid]:
            # 2 3 4 5 6 7 1
            if A[head] <= target and target <= A[mid]:
                tail = mid - 1
            else:
                head = mid + 1
        else:
            # A[head] > A[mid]
            # 5 6 7 1 2 3 4
            if A[mid] <= target and target <= A[tail]:
                head = mid + 1
            else:
                tail = mid - 1
    return -1
*/
